man/traq.1
.TH traq 1 "07 October 2013"

.SH NAME
bash based time tracking & evaluation

.SH SYNOPSIS
traq [options] [command or project]

.SH DESCRIPTION
A description of what the program does—as long as     is necessary.

.SH OPTIONS
.TP
.BI \-d " date"
print out stored informations for a given \fIdate\fR.
Can be combined with \fI\-p\fR to specify a single client.

.TP
.BI \-p " client"
track time for a given \fIclient\fR. Omit for default

.TP
.BI \-m " month"
print out stored informations for a given \fImonth\fR.
To specify a different year combine with \fI\-y\fR.

.TP
.BI \-y " year"
change the year used by \fI\-m\fR.

.SH EXAMPLES
To start tracking traq accepts a project you are working on:
.br
$ traq project

To stop tracking traq recognizes a special keyword called stop:
.br
$ traq stop

You can add comments to your tracked entries:
.br
$ traq project 'working on the landing page'

If you want to see which events you tracked today just call traq without arguments:
.br
$ traq

You can inspect dates/ weeks and months from the past:
.br
$ traq -d 2012-07-30
.br
$ traq -m 2

If you need to specify a different year
.br
$ traq -m 9 -y 2012

To tell traq to clock on different clients you can use \fI-p\fR:
.br
$ traq -p aClient development

Of course you can combine most options:
.br
$ traq -p aClient -y 2013 -m 02

.SH ENVIRONMENT
.TP
.BI TRAQ_PATH
the location where your traq installation resides. Defaults to $GOPATH/src/githum.com/nicolai86/traq
.TP
.BI TRAQ_DATA_DIR
the location where traq stores your files. Defaults to $HOME/Library/traq

.SH FILES
By default traq writes events to files located under $TRAQ_DATA_DIR. The directory needs to be read- and writeable for the current user. traq manages the data directory on its own.

.SH AUTHOR
Raphael Randschau <nicolai86@me.com>

CHANGELOG
# v0.6

- running timer evaluation using -r flag

# v0.5

- remove -w flag which printed times for a given week
- change directory structure of stored timestamps
- rewrite traq in go
- introduce -e flag to perform evaluations inside of traq

# v0.4

- automatically cancel trailing tags to ensure that traqeval does not calculate invalid
  durations

# v0.3

- traq now supports month and year batch printing of tracked times
  use -m to provide a month and -y to provide a year.
  Both -m and -y have support for bash completion, based on the current year
  and month.

# v0.2.1

- traqtrans works with pipes or file as first argument
- add tests for traqtrans and traqeval

# v0.2

- add support for comments

# v0.1

- add bash completion support
- add customizable data directory (@passy)

# v0.0.2

- traqeval is based on bash now
- add tests for traq_trans

# v0.0.1

- add support to track times using a simple text based format
- add support to track times for different projects using the `-p <project>` switch
- add support to list the current days tracking using an empty argument list
- add support to list the times for a week using the `-w <weeknumber>` switch
- add `traqtrans` programm to convert ISO 8601 timestamps to unix timestamps
- add `traqeval` programm to sum unix timestamps

traq.go
/*
Package traq implements helper methods for time tracking.
*/
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"regexp"
	"strings"
	"time"
)

// FilePath returns the path to a traq tracking file, taking the current
// env into account.
func FilePath(project string, date time.Time) (path string) {
	return fmt.Sprintf("%s/%s/%d/%d-%02d-%02d", os.Getenv("TRAQ_DATA_DIR"), project, date.Year(), date.Year(), date.Month(), date.Day())
}

// DatesInMonth calculates the days of a given month and year.
func DatesInMonth(year int, month int) []time.Time {
	var dates []time.Time = make([]time.Time, 0)
	var date time.Time = time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.UTC)

	for {
		dates = append(dates, date)
		date = date.Add(time.Hour * 24)
		if int(date.Month()) != month {
			break
		}
	}

	return dates
}

type LogLoader func(string) ([]string, error)

func ContentLoader(filePath string) ([]string, error) {
	content, err := ioutil.ReadFile(filePath)
	return strings.Split(string(content), "\n"), err
}

var stopLine = regexp.MustCompile(`;stop;`)
func RunningLoader(filePath string) ([]string, error) {
	content, err := ContentLoader(filePath)

	if err == nil {
		if stopLine.MatchString(content[len(content)-1]) {
			return content, err
		}

		var line = Entry(time.Now(), "stop")
		n := len(content)
		newContent := make([]string, n+1)
		copy(newContent, content)
		newContent[n] = line

		return newContent, err
	}

	return content, err
}

// SumFile evaluates the content of a traq tracking file.
// The returned map contains every tag contained in the file as well as the
// tracked duration in seconds.
func SumFile(lines []string) (map[string]int64, error) {
	var totalled map[string]int64 = make(map[string]int64)

	var currentTag string = ""
	var currentTime time.Time

	for _, line := range lines {
		if line != "" {
			var parts []string = strings.Split(line, ";")

			var t, error = time.Parse("Mon Jan 2 15:04:05 -0700 2006", parts[0])
			if error == nil {
				if parts[1] == "" {
					currentTag = parts[1]
					totalled[currentTag] = 0
				} else if parts[1] == "stop" {
					var diff = t.Unix() - currentTime.Unix()
					totalled[currentTag] = totalled[currentTag] + diff
					currentTag = ""
				} else if currentTag != parts[1] {
					var diff = t.Unix() - currentTime.Unix()
					totalled[currentTag] = totalled[currentTag] + diff
					currentTag = parts[1]
				}

				currentTime = t
			} else {
				return totalled, error
			}
		}
	}

	delete(totalled, "")

	return totalled, nil
}

// PrintDate prints the content of a single traqfile, identified by the project identifer
// and its date
func PrintDate(project string, dates ...time.Time) {
	for _, date := range dates {
		content, err := ioutil.ReadFile(FilePath(project, date))

		if err == nil {
			fmt.Print(string(content))
			fmt.Println("%%")
		}
	}
}

// EvaluateDate prints the evaluation of a single traqfile, identified by the project identifier
// and its date
func EvaluateDate(contentLoader LogLoader, project string, dates ...time.Time) {
	for _, date := range dates {
		var content, error = contentLoader(FilePath(project, date))

		if error == nil {
			fmt.Printf("%d-%02d-%02d\n", date.Year(), date.Month(), date.Day())
			var totalled, _ = SumFile(content)
			// TODO handle errors
			for key, value := range totalled {
				fmt.Printf("%s:%2.4f\n", key, float64(value)/60.0/60.0)
			}

			fmt.Println("%%")
		}
	}
}

func Entry(date time.Time, command string) string {
	return fmt.Sprintf("%s;%s;%s\n", date.Format("Mon Jan 2 15:04:05 -0700 2006"), command, "")
}

// WriteToFile writes a given command to a traq file, converting it into a tag
// if it's no known command.
func WriteToFile(project string, date time.Time, command string) {
	if command != "stop" {
		command = "#" + command
	}

	var traqFile string = FilePath(project, date)
	var projectDir string = path.Dir(traqFile)

	_ = os.MkdirAll(projectDir, os.ModeDir|os.ModePerm)
	var file, error = os.OpenFile(traqFile, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0660)
	if error == nil {
		file.WriteString(Entry(date, command))
		file.Close()
	}
}

var month int
var year int
var day int
var project string = "timestamps"
var date string
var evaluate bool
var running bool

func main() {
	flag.BoolVar(&evaluate, "e", false, "evaluate tracked times")
	flag.BoolVar(&running, "r", false, "add fake stop entry to evaluate if stop is missing")

	flag.IntVar(&year, "y", 0, "print tracked times for a given year")
	flag.IntVar(&month, "m", 0, "print tracked times for a given month")

	flag.StringVar(&date, "d", "", "print tracked times for a given date")
	flag.StringVar(&project, "p", "", "print data for a given project")

	flag.Parse()

	var now = time.Now()
	var t, error = time.Parse("2006-01-02", date)
	if error == nil {
		year = t.Year()
		month = int(t.Month())
		day = t.Day()
	} else {
		if month == 0 && year == 0 {
			day = now.Day()
		} else {
			day = 1
		}
		if year == 0 {
			year = now.Year()
		}
		if month == 0 {
			month = int(now.Month())
		}
	}

	var loader LogLoader = ContentLoader

	if running {
		loader = RunningLoader
	}

	if evaluate {
		if date == "" {
			EvaluateDate(loader, project, DatesInMonth(year, month)...)
		} else {
			EvaluateDate(loader, project, t)
		}
		return
	}

	var command string = flag.Arg(0)

	if command == "" {
		if date == "" {
			PrintDate(project, DatesInMonth(year, month)...)
		} else {
			PrintDate(project, t)
		}
	} else {
		WriteToFile(project, now, command)
	}
}
